File: src/geek time/c#/week03/95. Unique Binary Search Trees II.cs
using System;
using System.Collections.Generic;

namespace week03 {
    public partial class Solution {
        public IList<TreeNode> GenerateTrees(int n) {
            // 这个题自己实现的时候就存在一个问题，不知道递归应该生成的内容是什么
            // 需要注意：这种生成树的题，递归for循环应该写在递归体内，因为这样才能不断的递归生成树
            if (n == 0) {
                return new List<TreeNode>();
            }
            return RecursiveGenerate(1, n);
        }

        private List<TreeNode> RecursiveGenerate(int start, int end) {
            var list = new List<TreeNode>();
            if (start > end) {
                list.Add(null);
                return list;
            }
            for (var i = start; i <= end; i++) {
                var leftTrees = RecursiveGenerate(start, i - 1);
                var rightTrees = RecursiveGenerate(i + 1, end);
                foreach(var left in leftTrees) {
                    foreach(var right in rightTrees) {
                        var tree = new TreeNode(i);
                        tree.left = left;
                        tree.right = right;
                        list.Add(tree);
                    }
                }
            }
            return list;
        }
    }
}

File: src/geek time/typescript/week03/95. Unique Binary Search Trees II.ts
export class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

export function generateTrees(n: number): Array<TreeNode | null> {
    if (!n) return [];
    return recursiveGenerate(1, n);
}
function recursiveGenerate(start: number, end: number): TreeNode[] {
    const ans: TreeNode[] = [];
    if (start > end) {
        ans.push(null);
        return ans;
    }
    for (let i = start; i <= end; i++) {
        const leftTrees = recursiveGenerate(start, i - 1);
        const rightTrees = recursiveGenerate(i + 1, end);
        leftTrees.forEach((left) => {
            rightTrees.forEach((right) => {
                const tree = new TreeNode(i);
                tree.left = left;
                tree.right = right;
                ans.push(tree);
            });
        });
    }
    return ans;
}
